ambry-store/src/main/java/com/github/ambry/store/BlobStoreCompactor.java

```diff
@@ -1467,7 +1485,15 @@ public boolean alreadyExists(PersistentIndex idx, FileSpan searchSpan, StoreKey
                 "Delete's latest state can't be put for key" + currentKey + " in store " + dataDir);
           }
           if (currentLatestState.isDelete() && currentLatestState.getLifeVersion() == currentValue.getLifeVersion()) {
-            validEntries.add(entry);
+            // check if this is a removable delete tombstone.
+            if (config.storeCompactionPurgeExpiredDeleteTombstone && isDeleteTombstoneRemovable(currentValue)
+                && getPutValueFromSrc(currentKey, currentValue, indexSegment) == null) {
```

Comment:
We feel it can be removed as well,  but we are not fully sure all edge cases have been covered.  Leaving this condition here is a more conservative way to remove delete tombstone without changing existing compaction logic. We can revisit this once we have more confidence to clean up PUT and DELETE together.